File: 2016/app/src/main/java/aoc2016/solutions/Day24.java
package aoc2016.solutions;

import java.util.ArrayList;

import aoc2016.utils.FileLoader;
import aoc2016.utils.Point;

public class Day24 {
    public static void solve() {
        String input = FileLoader.loadFile("day24.txt");
        System.out.println("Day 24 part 1: " + part1(input));
        System.out.println("Day 24 part 2: " + part2(input));
    }

    private static int part1(String input) {
        Grid grid = new Grid(input);
        // Print POIs
        for (Point poi : grid.pois) {
            System.out.println(poi);
        }
        return 0;
    }

    private static int part2(String input) {
        return 0;
    }

    private static class Grid {
        ArrayList<ArrayList<Character>> grid;
        ArrayList<Point> pois;

        Grid(String input) {
            grid = new ArrayList<>();
            pois = new ArrayList<>();
            for (String line : input.split("\n")) {
                ArrayList<Character> row = new ArrayList<>();
                for (char c : line.toCharArray()) {
                    row.add(c);
                }
                grid.add(row);
            }
            for (int y = 0; y < grid.size(); y++) {
                for (int x = 0; x < grid.get(y).size(); x++) {
                    if (grid.get(y).get(x) != '#' && grid.get(y).get(x) != '.') {
                        pois.add(new Point(x, y));
                    }
                }
            }
        }
    }
}

File: 2017/app/src/main/kotlin/aoc/solutions/Day24.kt
package aoc.solutions

import aoc.utils.FileLoader

fun solveDayTwentyFour() {
    val input = FileLoader("day24.txt").readLines()
    println("Day 24 part 1: ${partOne(input)}")
    println("Day 24 part 2: ${partTwo(input)}")
}

private fun partTwo(input: List<String>): Int {
    val components = input.map { Component.fromLine(it) }
    val bridges = mutableListOf<List<Component>>()
    val used = mutableSetOf<Component>()
    buildBridges(0, components, bridges, used)
    val max = bridges.maxOf { it.size }
    return bridges.filter { it.size == max }.maxOf { it.sumOf { it.a + it.b } }
}

private fun partOne(input: List<String>): Int {
    val components = input.map { Component.fromLine(it) }
    val bridges = mutableListOf<List<Component>>()
    val used = mutableSetOf<Component>()
    buildBridges(0, components, bridges, used)
    return bridges.maxOf { it.sumOf { it.a + it.b } }
}

private fun buildBridges(
    port: Int,
    components: List<Component>,
    bridges: MutableList<List<Component>>,
    used: MutableSet<Component>,
    currBridge: List<Component> = listOf(),
) {
    val possibleComponents = components.filter { it.a == port || it.b == port }
    for (component in possibleComponents) {
        if (component in used) {
            continue
        }
        val newBridge = currBridge + component
        used.add(component)
        buildBridges(if (component.a == port) component.b else component.a, components, bridges, used, newBridge)
        used.remove(component)
    }

    bridges.add(currBridge)
}

private data class Component(val a: Int, val b: Int) {
    companion object {
        fun fromLine(line: String): Component {
            val (a, b) = line.split("/").map { it.toInt() }
            return Component(a, b)
        }
    }
}
